src/main/java/org/opensearch/geospatial/ip2geo/action/PutDatasourceTransportAction.java

```diff
@@ -0,0 +1,189 @@
+import org.opensearch.geospatial.ip2geo.jobscheduler.Datasource;
+import org.opensearch.geospatial.ip2geo.jobscheduler.DatasourceExtension;
+import org.opensearch.index.engine.VersionConflictEngineException;
+import org.opensearch.tasks.Task;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportService;
+
+/**
+ * Transport action to create datasource
+ */
+@Log4j2
+public class PutDatasourceTransportAction extends HandledTransportAction<PutDatasourceRequest, AcknowledgedResponse> {
+    private final Client client;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
+    private TimeValue timeout;
+    private int indexingBulkSize;
+
+    /**
+     * Default constructor
+     * @param transportService the transport service
+     * @param actionFilters the action filters
+     * @param client the client
+     * @param clusterService the cluster service
+     * @param threadPool the thread pool
+     * @param settings the settings
+     * @param clusterSettings the cluster settings
+     */
+    @Inject
+    public PutDatasourceTransportAction(
+        final TransportService transportService,
+        final ActionFilters actionFilters,
+        final Client client,
+        final ClusterService clusterService,
+        final ThreadPool threadPool,
+        final Settings settings,
+        final ClusterSettings clusterSettings
+    ) {
+        super(PutDatasourceAction.NAME, transportService, actionFilters, PutDatasourceRequest::new);
+        this.client = client;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        timeout = Ip2GeoSettings.TIMEOUT_IN_SECONDS.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(Ip2GeoSettings.TIMEOUT_IN_SECONDS, newValue -> timeout = newValue);
+        indexingBulkSize = Ip2GeoSettings.INDEXING_BULK_SIZE.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(Ip2GeoSettings.INDEXING_BULK_SIZE, newValue -> indexingBulkSize = newValue);
+    }
+
+    @Override
+    protected void doExecute(final Task task, final PutDatasourceRequest request, final ActionListener<AcknowledgedResponse> listener) {
+        try {
+            Datasource jobParameter = Datasource.Builder.build(request);
+            IndexRequest indexRequest = new IndexRequest().index(DatasourceExtension.JOB_INDEX_NAME)
+                .id(jobParameter.getId())
+                .source(jobParameter.toXContent(JsonXContent.contentBuilder(), null))
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .opType(DocWriteRequest.OpType.CREATE);
+            client.index(indexRequest, new ActionListener<>() {
+                @Override
+                public void onResponse(final IndexResponse indexResponse) {
+                    // This is user initiated request. Therefore, we want to handle the first datasource update task in a generic thread
+                    // pool.
```

Comment:
qq: Could you explain this a little bit more?

Reply:
There are two flow where geoip data is created. 1. when user call create datasource api. 2. auto update after specified interval. For the second one, we are using special thread pool for Ip2Geo which has very limited resource like single thread to minimized the impact to the cluster availability. If we use the thread pool for Ip2Geo for the first case, the very first datasource creation will get delayed when there are another update going on. Therefore, I choose to use generic thread pool for the first case.

Reply:
Why would another update be going on? Im not sure we should use a different threadpool for initial vs. update.

Reply:
There can be multiple datasource. One of them might be running auto update process. During that time, a user can call a create another datasource request. If we use same thread pool, this initial creation will get delayed as it cannot start before the ongoing update finishes. 
Also, it might get more difficult to handle race condition between multiple request into same datasource if the first request gets delayed to be handled.